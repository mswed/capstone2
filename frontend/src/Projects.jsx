import { useState, useEffect } from 'react';
import { Container, Card, Row, Col } from 'react-bootstrap';
import GrumpyApi from './api';
import Loading from './Loading';
import RemoteSearchForm from './RemoteSearchForm';
import ProjectList from './ProjectList';
import NoResults from './NoResults';

const Projects = () => {
  const [projects, setProjects] = useState({ local: [], remote: [] });
  const [searchTerm, setSearchTerm] = useState('');
  const [isLoading, setIsLoading] = useState(true);

  // Conditions for rendering
  const hasSearch = searchTerm.trim() !== '';
  const hasLocalResults = projects.local.length > 0;
  const hasRemoteResults = projects.remote.length > 0;

  /**
   * Get all the projects in the database. Eventually this should have a limit
   * or pagination but for now we get everything. This is called by useEffect
   * if the searchTerm changes and is empty
   *
   * @returns {Array} all projects in the database
   */

  const getAllProjects = async () => {
    try {
      const response = await GrumpyApi.getProjects();
      setProjects({ local: response, remote: [] });
    } catch (error) {
      console.error('Error fetching cameras', error);
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Get found projects in the databse. This is called by useEffect
   * if the searchTerm changes and is NOT empty
   *
   * @returns {Array} found projects in the database
   */

  const findProjects = async () => {
    try {
      const response = await GrumpyApi.findProjects(searchTerm);
      setProjects(response);
      console.log('search set state to', projects);
    } catch (error) {
      console.error('Error fetching projects', error);
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Add a project from TMDB to the local database
   *
   * @param {integer} projectId - Project id from TMDB
   * @param {string} projectType - Project type (episodic or feature)
   */

  const addProject = async (projectId, projectType) => {
    const newProject = await GrumpyApi.addTMDBProject(projectId, projectType);
    setProjects((prev) => ({
      ...prev,
      local: [...prev.local, newProject.project],
    }));
  };

  useEffect(() => {
    if (searchTerm.trim() !== '') {
      findProjects();
    } else {
      getAllProjects();
    }
  }, [searchTerm]);

  if (isLoading) {
    return <Loading />;
  }
  return (
    <Container>
      <RemoteSearchForm search={setSearchTerm} />
      <div className="mb-4">
        <h3 className="text-start">In Database</h3>
        {hasLocalResults ? (
          <ProjectList projects={projects.local} />
        ) : hasSearch ? (
          <NoResults message="No projects found in the databse" />
        ) : (
          <NoResults message="Please search for projects" />
        )}
      </div>
      {hasSearch && (
        <div>
          <h3 className="text-start">Browse TMDB</h3>{' '}
          {hasRemoteResults ? <ProjectList projects={projects.remote} projectsType="remote" onProjectAdd={addProject} /> : <NoResults message="No results found on TMDB" />}
        </div>
      )}
    </Container>
  );
};

export default Projects;
