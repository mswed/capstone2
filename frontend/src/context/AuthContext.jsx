import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
