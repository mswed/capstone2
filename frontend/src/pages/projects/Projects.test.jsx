import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor, within } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Projects from './Projects.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';

// Formats search scrolls up when results change so we need to mock
// scrollTo
global.scrollTo = vi.fn();

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getProjects: vi.fn(() =>
      Promise.resolve([
        {
          id: 1,
          name: 'One Piece',
          projectType: 'episodic',
          description: 'With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure.',
          poster_path: 'https://www.pathtoposter1.com',
          releaseDate: '2023-08-31',
          adult: false,
          tmdb_id: 123456,
          tmdbOriginalName: 'same as the other name',
          genres: ['adventure'],
          rating: 'PG-13',
        },

        {
          id: 2,
          name: 'Daredevil',
          projectType: 'feature',
          description: 'A pretty bad movie',
          poster_path: 'https://www.pathtoposter2.com',
          releaseDate: '2010-02-01',
          adult: false,
          tmdb_id: 654321,
          tmdbOriginalName: 'Daredevil is silly',
          genres: ['action', 'adventure'],
          rating: 'R',
        },
      ])
    ),
    findProjects: vi.fn(() =>
      Promise.resolve({
        projects: {
          local: [
            {
              id: 1,
              name: 'One Piece But it is local',
              projectType: 'episodic',
              description: 'A goofy pirate adventure',
              poster_path: 'https://www.pathtoposter1.com',
              releaseDate: '2024-01-01',
              adult: false,
              tmdb_id: 123456,
              tmdbOriginalName: 'same as the other name',
              genres: ['adventure'],
              rating: 'PG-13',
            },
          ],
          remote: [
            {
              id: 2,
              name: 'Daredevil but it is remote',
              projectType: 'feature',
              description: 'A pretty bad movie',
              poster_path: 'https://www.pathtoposter2.com',
              releaseDate: '2010-02-01',
              adult: false,
              tmdb_id: 654321,
              tmdbOriginalName: 'Daredevil is silly',
              genres: ['action', 'adventure'],
              rating: 'R',
            },
          ],
        },
      })
    ),
  },
}));

describe('Projects component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter>
        <AuthContext.Provider value={authValue}>
          <Projects />
        </AuthContext.Provider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findAllByText('One Piece');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findAllByText('One Piece');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();
    await screen.findAllByText('In Database');

    const table = screen.getByRole('table');
    const rows = within(table).getAllByRole('row');

    await waitFor(() => {
      // Table headers
      expect(screen.getByRole('columnheader', { name: 'Image' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Name' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Type' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Release Date' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Description' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Action' })).toBeInTheDocument();

      // Table data
      const firstDataRow = rows[1]; // Skip header row
      expect(within(firstDataRow).getByText('One Piece')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('2023-08-31')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('episodic')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure.')).toBeInTheDocument();
    });
  });

  it('can search for projets', async () => {
    renderWithAuth();

    await screen.findAllByText('In Database');

    const searchInput = screen.getByPlaceholderText('Enter search term...');
    fireEvent.change(searchInput, { target: { value: 'test search' } });
    const searchButton = screen.getByRole('button', { name: 'Search!' });

    fireEvent.click(searchButton);

    await waitFor(() => {
      // Find the sections by their headings
      const localSection = screen.getByText('In Database').closest('div');
      const remoteSection = screen.getByText('Browse TMDB').closest('div');

      expect(within(localSection).getByText('One Piece But it is local')).toBeInTheDocument();
      expect(within(remoteSection).getByText('Daredevil but it is remote')).toBeInTheDocument();
    });
  });
});
