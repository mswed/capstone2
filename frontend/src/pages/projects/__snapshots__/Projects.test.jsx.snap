// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Projects component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class=""
    >
      <div
        class="my-3 input-group"
      >
        <input
          class="form-control"
          placeholder="Enter search term..."
          value=""
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search!
        </button>
      </div>
    </form>
    <div
      class="mb-4"
    >
      <h3
        class="text-start"
      >
        In Database
      </h3>
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-hover"
        >
          <thead>
            <tr>
              <th>
                Image
              </th>
              <th>
                Name
              </th>
              <th>
                Release Date
              </th>
              <th>
                Type
              </th>
              <th>
                Description
              </th>
              <th>
                Action
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <img
                  alt="Image of undefined"
                  src="/media/camera_images/missing_image.png"
                  style="width: 60px;"
                />
              </td>
              <td>
                One Piece
              </td>
              <td>
                2023-08-31
              </td>
              <td>
                episodic
              </td>
              <td
                class="text-start"
              >
                With his straw hat and ragtag crew, young pirate Monkey D. Luffy goes on an epic voyage for treasure.
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/projects/1"
                >
                  View
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Image of undefined"
                  src="/media/camera_images/missing_image.png"
                  style="width: 60px;"
                />
              </td>
              <td>
                Daredevil
              </td>
              <td>
                2010-02-01
              </td>
              <td>
                feature
              </td>
              <td
                class="text-start"
              >
                A pretty bad movie
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/projects/2"
                >
                  View
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
